pymc_marketing/mmm/utility.py

```diff
@@ -0,0 +1,546 @@
+#   Copyright 2024 The PyMC Labs Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+#   Copyright 2024 The PyMC Labs Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""
+Utility functions for Bayesian optimization.
+
+Key Concepts:
+-------------
+
+- **Samples**:
+    A PyTensor tensor variable (`pt.TensorVariable`) representing samples drawn from the posterior
+    distributions of the model outputs. These samples capture the uncertainty in the model predictions
+    and are essential for computing expected utilities and risk measures in Bayesian optimization.
+
+- **Budgets**:
+    A PyTensor tensor variable representing a set of monetary budgets allocated to different assets,
+    investments, or channels. Each element corresponds to the budget for a specific option in the
+    optimization process.
+"""
+
+from collections.abc import Callable
+
+import pytensor.tensor as pt
+
+UtilityFunction = Callable[[pt.TensorVariable, pt.TensorVariable], float]
```

Comment:
Shall we call this `UtilityFunctionType`? Otherwise it could be interpreted as a class

Reply:
Yes, we can do that!